File: frontend/src/services/websocket.js
/**
 * WebSocket Client for Real-Time Audio Streaming
 * Uses Socket.IO for bidirectional communication
 */

const WS_URL = process.env.REACT_APP_WS_URL || 'ws://localhost:5000';

class WebSocketClient {
    constructor(url) {
        this.url = url;
        this.socket = null;
        this.sessionId = null;
        this.callbacks = {
            connected: [],
            disconnected: [],
            processed_audio: [],
            metrics_update: [],
            error: []
        };
    }

    /**
     * Connect to WebSocket server
     */
    connect() {
        if (this.socket && this.socket.connected) {
            console.log('Already connected');
            return;
        }

        this.socket = io(this.url, {
            transports: ['websocket', 'polling'],
            reconnection: true,
            reconnectionDelay: 1000,
            reconnectionAttempts: 5
        });

        // Setup event listeners
        this.socket.on('connect', () => {
            console.log('WebSocket connected');
            this._trigger('connected');
        });

        this.socket.on('disconnect', () => {
            console.log('WebSocket disconnected');
            this._trigger('disconnected');
        });

        this.socket.on('processed_audio', (data) => {
            this._trigger('processed_audio', data);
        });

        this.socket.on('metrics_update', (data) => {
            this._trigger('metrics_update', data);
        });

        this.socket.on('error', (error) => {
            console.error('WebSocket error:', error);
            this._trigger('error', error);
        });
    }

    /**
     * Disconnect from server
     */
    disconnect() {
        if (this.socket) {
            this.socket.disconnect();
            this.socket = null;
        }
    }

    /**
     * Join audio processing session
     */
    joinSession(sessionId) {
        this.sessionId = sessionId;
        this.socket.emit('join_session', { session_id: sessionId });
    }

    /**
     * Leave current session
     */
    leaveSession() {
        if (this.sessionId) {
            this.socket.emit('leave_session', { session_id: this.sessionId });
            this.sessionId = null;
        }
    }

    /**
     * Send audio chunk for processing
     */
    sendAudioChunk(audioData, options = {}) {
        if (!this.sessionId) {
            console.error('No active session');
            return;
        }

        this.socket.emit('audio_chunk', {
            session_id: this.sessionId,
            audio_data: audioData,
            sample_rate: options.sample_rate || 48000,
            algorithm: options.algorithm || 'nlms',
            intensity: options.intensity || 1.0,
            chunk_index: options.chunk_index || 0
        });
    }

    /**
     * Request metrics update
     */
    requestMetrics() {
        if (!this.sessionId) {
            console.error('No active session');
            return;
        }

        this.socket.emit('request_metrics', {
            session_id: this.sessionId
        });
    }

    /**
     * Register event callback
     */
    on(event, callback) {
        if (this.callbacks[event]) {
            this.callbacks[event].push(callback);
        }
    }

    /**
     * Unregister event callback
     */
    off(event, callback) {
        if (this.callbacks[event]) {
            this.callbacks[event] = this.callbacks[event].filter(cb => cb !== callback);
        }
    }

    /**
     * Trigger event callbacks
     */
    _trigger(event, data) {
        if (this.callbacks[event]) {
            this.callbacks[event].forEach(callback => callback(data));
        }
    }

    /**
     * Check if connected
     */
    isConnected() {
        return this.socket && this.socket.connected;
    }
}

// Export singleton instance
const wsClient = new WebSocketClient(WS_URL);

// For vanilla JS
if (typeof window !== 'undefined') {
    window.wsClient = wsClient;
}

// For ES modules
export default wsClient;

File: backend/websocket.py
"""
WebSocket Server for Real-Time Audio Streaming
"""

from flask_socketio import emit, join_room, leave_room
import logging

logger = logging.getLogger(__name__)


def init_socketio(socketio):
    """Initialize WebSocket event handlers"""

    @socketio.on('connect')
    def handle_connect():
        """Handle client connection"""
        logger.info(f"Client connected")
        emit('connected', {'message': 'Connected to ANC server'})

    @socketio.on('disconnect')
    def handle_disconnect():
        """Handle client disconnection"""
        logger.info(f"Client disconnected")

    @socketio.on('join_session')
    def handle_join_session(data):
        """Join audio processing session"""
        session_id = data.get('session_id')
        if session_id:
            join_room(session_id)
            logger.info(f"Client joined session: {session_id}")
            emit('session_joined', {'session_id': session_id})

    @socketio.on('leave_session')
    def handle_leave_session(data):
        """Leave audio processing session"""
        session_id = data.get('session_id')
        if session_id:
            leave_room(session_id)
            logger.info(f"Client left session: {session_id}")
            emit('session_left', {'session_id': session_id})

    @socketio.on('audio_chunk')
    def handle_audio_chunk(data):
        """
        Handle incoming audio chunk for processing

        Data format:
        {
            'session_id': 'uuid',
            'audio_data': 'base64-encoded-audio',
            'sample_rate': 48000,
            'chunk_index': 0
        }
        """
        try:
            session_id = data.get('session_id')
            audio_data = data.get('audio_data')
            sample_rate = data.get('sample_rate', 48000)

            # Process audio with ANC
            from services.anc_service import ANCService
            anc_service = ANCService()

            result = anc_service.process_audio(
                audio_data=audio_data,
                sample_rate=sample_rate,
                algorithm=data.get('algorithm', 'nlms'),
                intensity=data.get('intensity', 1.0)
            )

            # Send processed audio back
            emit('processed_audio', {
                'session_id': session_id,
                'processed_audio': result['processed_audio'],
                'metrics': result['metrics'],
                'chunk_index': data.get('chunk_index')
            }, room=session_id)

            logger.debug(f"Processed audio chunk for session: {session_id}")

        except Exception as e:
            logger.error(f"Error processing audio chunk: {e}")
            emit('error', {'message': str(e)})

    @socketio.on('request_metrics')
    def handle_request_metrics(data):
        """Send real-time metrics update"""
        session_id = data.get('session_id')

        # Get session metrics from database
        try:
            from database.models import AudioSession

            session = AudioSession.query.get(session_id)
            if session:
                emit('metrics_update', {
                    'session_id': session_id,
                    'metrics': {
                        'chunks_processed': session.total_chunks_processed,
                        'avg_latency_ms': session.average_latency_ms,
                        'avg_cancellation_db': session.average_cancellation_db
                    }
                }, room=session_id)

        except Exception as e:
            logger.error(f"Error fetching metrics: {e}")

    logger.info("WebSocket event handlers initialized")
